feat: Install plugins using schedule name

pnadolny13 opened the pull request:
Related to https://github.com/meltano/meltano/issues/7001

Adds the ability to install plugins in a given schedule, helpful for workers that dont have preinstalled plugins. Personally this would be beneficial to me because it would eliminate the need for my CI install script that attempts to do the same thing [ci_get_install_script.py](https://github.com/meltano/squared/blob/main/data/ci_get_install_script.py) and also speeds up my cloud schedule startup time. I have 34 plugins to install before every cloud run so for a bunch of my schedules the install time is the same or more than my actual execution time and I think this would cut it down significantly.

Notes:
- The logic is added after the normal plugin type filtering so you can still filter for "extractors in schedule xyz"
- I confirmed that this works locally but when I went to write unit tests I wasnt quite sure how this fits in. Before I spend time on that I figured I'd get some feedback if this logic is better suited somewhere else (e.g. something along the lines of a get_job_plugins in ScheduleService or TaskSetService).

pnadolny13 committed:
install using schedule name

pnadolny13 committed:
revert else change

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.ci hooks

for more information, see https://pre-commit.ci

tayloramurphy commented on src/meltano/cli/install.py:
```diff
@@ -44,6 +45,11 @@
     help="Ignore the required Python version declared by the plugins.",
 )
+@click.option(
+    "--schedule_name",
```
Very cool PR @pnadolny13 ! Add a docs update around this as well too please 😄

pnadolny13 commented:
@WillDaSilva @kgpayne @cjohnhanson any thoughts on whether this is the wrong place for this? I'm happy to make any changes and add tests but didnt want to spend too much time until I heard your thoughts since I'm not super familiar with the codebase.

WillDaSilva reviewed (commented):
@pnadolny13 This feature is great! A simple way for us to massively improve Meltano Cloud performance before we get around to creating images for each deployment, or cached plugin venvs/zipapps.

> any thoughts on whether this is the wrong place for this?

Ideally we wouldn't have any of this application logic in the CLI function, but it's already the case that many of the Meltano CLI functions have lots of application logic, and that's already the case in this specific CLI function, so I don't really care right now. Holding this improvement back so that we can refactor & clean up the CLI is not worth it, and it's not like this feature would make it more complicated to perform that refactoring / clean up later either.

pnadolny13 committed:
add unit test for schedule install

pnadolny13 committed:
add docs for schedule install

pnadolny13 committed:
Merge branch 'main' into install_via_schedule

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.ci hooks

for more information, see https://pre-commit.ci

pnadolny13 commented:
@WillDaSilva thanks for the guidance. I updated this PR with a passing unit test, its kind of ugly because I feel like I'm forcing this code block into the CLI function but it does the job. I also added a few lines to the cli reference docs.

WillDaSilva commented on tests/meltano/cli/test_install.py:
```diff
@@ -236,6 +236,42 @@ def test_clean_install(
+            )
+            result = cli_runner.invoke(
+                cli,
+                ["install", "--schedule_name", job_schedule.name],
```
```suggestion
                ["install", "--schedule", job_schedule.name],
```

pnadolny13 commented on docs/src/_reference/command-line-interface.md:
```diff
@@ -628,6 +628,9 @@ Installs dependencies of your project based on the **meltano.yml** file.
 Optionally, provide a plugin type argument to only (re)install plugins of a certain type.
 Additionally, plugin names can be provided to only (re)install those specific plugins.
 
+To only install plugins for a particular schedule specify the `--schedule_name` argument.
```
```suggestion
To only install plugins for a particular schedule specify the `--schedule` argument.
```

pnadolny13 committed:
Apply suggestions from code review

Co-authored-by: Will Da Silva <will@willdasilva.xyz>

pnadolny13 committed:
test has unordered assert, so limit to just dbt plugin

pnadolny13 committed:
reduce complexity by pulling logic out to standalone method

pnadolny13 committed:
Update src/meltano/cli/install.py

Co-authored-by: Will Da Silva <will@willdasilva.xyz>

pnadolny13 committed:
Merge branch 'install_via_schedule' of https://github.com/meltano/meltano into install_via_schedule

pnadolny13 committed:
Merge branch 'main' into install_via_schedule

pnadolny13 committed:
Merge branch 'install_via_schedule' of https://github.com/meltano/meltano into install_via_schedule
